### PR TITLE
checks: Add Image Check, context object to Run

### PR DESCRIPTION
**Pull request by nathanjsweet**
1. Change Run interface to accept context parameter, so run can receive CLI args.
2. Add Image check using docker-reference lib.
3. Add unit tests for image checks.

### CODE REVIEW

File: checks/image/image_check.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+func init() {
+	checks.Register(&containerImageCheck{})
+}
+
+type containerImageCheck struct{}
+
+// Name returns a unique name for this check.
+func (pc *containerImageCheck) Name() string {
+	return "container-images"
```
Comment (varshavaradarajan) [inline checks/image/image_check.go]: Can this be more specific? There are a couple of checks for container images.

File: checks/image/image_check.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	return diagnostics, nil
+}
+
+// checkPod checks if all pod images are well formed and
+// do not have any errors, or obvious mistakes.
+func (pc *containerImageCheck) checkPod(validRepos map[string]struct{}, pod corev1.Pod) []checks.Diagnostic {
+	var diagnostics []checks.Diagnostic
+	containers := append(pod.Spec.Containers, pod.Spec.InitContainers...)
```
Comment (varshavaradarajan) [inline checks/image/image_check.go]: What about init containers? Similar to https://github.com/digitalocean/clusterlint/blob/master/checks/basic/fully_qualified_image.go#L56

File: checks/image/image_check.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	errInvalidDomain = errors.New("invalid domain")
+	errNoSha         = errors.New("no sha")
+)
+
+func lintImage(named reference.Named, validRepos map[string]struct{}, containerName, image string) (warnings []error) {
+	domain := reference.Domain(named)
+	// the reference lib inserts "docker.io" into the image if it is not present,
+	// we remove it if we find that it isn't specified.
+	if domain == dockerIO && strings.Index(image, dockerIO) != 0 {
+		domain = ""
```
Comment (varshavaradarajan) [inline checks/image/image_check.go]: append to warnings here and don't check if domain is blank again?

File: checks/image/image_check.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+					Owners:   pod.ObjectMeta.GetOwnerReferences(),
+				})
+			}
+		}
+	}
+	return diagnostics
+}
+
+func checkImageError(containerName, image string) (reference.Named, error) {
+	named, err := reference.ParseNormalizedNamed(image)
```
Comment (varshavaradarajan) [inline checks/image/image_check.go]: We do a check for malformed image name in https://github.com/digitalocean/clusterlint/blob/master/checks/basic/fully_qualified_image.go. I wonder if it is necessary in both places actually. If there is an error due to malformed image or anything else, will the pod object even be available for us to check? https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#podlist-v1-core just says list of pods. I'll check and update here.

File: checks/image/image_check.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+type containerImageCheck struct{}
+
+// Name returns a unique name for this check.
+func (pc *containerImageCheck) Name() string {
+	return "container-images"
+}
+
+// Groups returns a list of group names this check should be part of.
+func (pc *containerImageCheck) Groups() []string {
+	return []string{"image"}
```
Comment (varshavaradarajan) [inline checks/image/image_check.go]: @adamwg - I like adding checks based on what object the check deals with. Maybe we should do that for all our checks and then allow users to run say just pod related checks or, node related checks.. WDYT? @nathanjsweet - how about changing this to pod then?

File: checks/image/image_check.go [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+// Description returns a detailed human-readable description of what this check
+// does.
+func (pc *containerImageCheck) Description() string {
+	return "Checks if there are pods with containers in privileged mode"
```
Comment (varshavaradarajan) [inline checks/image/image_check.go]: Change the desc?

File: checks/image/image_check.go (hunk 2 above)
Comment (nathanjsweet) [reply checks/image/image_check.go]: This does check both.

File: checks/image/image_check.go (hunk 4 above)
Comment (nathanjsweet) [reply checks/image/image_check.go]: Okay, this is interesting. These should definitely get moved there. Good catch. However, if it's okay with you and @adamwg  I think we should rework the logic there a little bit. The docker `reference` package of detecting errors in image names, which should be out-and-out cluster lint errors, IMO. Thereafter non-fully-qualified images and other warnings could be added. Thoughts?

File: checks/image/image_check.go (hunk 5 above)
Comment (nathanjsweet) [reply checks/image/image_check.go]: Yeah, that's a good idea.

File: checks/image/image_check.go (hunk 6 above)
Comment (nathanjsweet) [reply checks/image/image_check.go]: Good catch. Thanks!

File: checks/image/image_check.go (hunk 3 above)
Comment (nathanjsweet) [reply checks/image/image_check.go]: This is to set the domain to empty if "docker.io" is added by `ParseNormalizedNamed`.  Thereafter an empty check is made.

**Pull request closed without merging**